kernel/src/actions/mod.rs

```diff
@@ -321,42 +337,51 @@ impl Protocol {
     }
 }
 
-// given unparsed `table_features`, parse and check if they are subset of `supported_features`
+// given `table_features`, check if they are subset of `supported_features`
 pub(crate) fn ensure_supported_features<T>(
-    table_features: &[String],
-    supported_features: &HashSet<T>,
+    table_features: &[T],
+    supported_features: &[T],
 ) -> DeltaResult<()>
 where
+    T: Display + FromStr + Hash + Eq,
     <T as FromStr>::Err: Display,
-    T: Debug + FromStr + Hash + Eq,
 {
-    let error = |unsupported, unsupported_or_unknown| {
-        let supported = supported_features.iter().collect::<Vec<_>>();
-        let features_type = type_name::<T>()
-            .rsplit("::")
-            .next()
-            .unwrap_or("table features");
-        Error::Unsupported(format!(
-            "{} {} {:?}. Supported {} are {:?}",
-            unsupported_or_unknown, features_type, unsupported, features_type, supported
-        ))
-    };
-    let parsed_features: HashSet<T> = table_features
+    // first check if all features are supported, else we proceed to craft an error message
+    if table_features
         .iter()
-        .map(|s| T::from_str(s).map_err(|_| error(vec![s.to_string()], "Unknown")))
-        .collect::<Result<_, Error>>()?;
-
-    // check that parsed features are a subset of supported features
-    parsed_features
-        .is_subset(supported_features)
-        .then_some(())
-        .ok_or_else(|| {
-            let unsupported = parsed_features
-                .difference(supported_features)
-                .map(|f| format!("{:?}", f))
-                .collect::<Vec<_>>();
-            error(unsupported, "Unsupported")
-        })
+        .all(|feature| supported_features.contains(feature))
+    {
+        return Ok(());
+    }
+
+    // we get the type name (ReaderFeature/WriterFeature) for better error messages
+    let features_type = std::any::type_name::<T>()
+        .rsplit("::")
+        .next()
+        .unwrap_or("table feature");
+
+    // NB: we didn't do this above to avoid allocation in the common case
+    let mut unsupported: Vec<_> = table_features
+        .iter()
+        .filter(|feature| !supported_features.contains(*feature))
+        .map(|feature| feature.to_string())
+        .collect();
+    let mut supported: Vec<_> = supported_features
+        .iter()
+        .map(|feature| feature.to_string())
+        .collect();
+
+    // sort both vecs for consistent error messages
+    unsupported.sort();
+    supported.sort();
+
+    Err(Error::Unsupported(format!(
+        "Unknown {}s: \"{}\". Supported {}s: \"{}\"",
+        features_type,
+        unsupported.join("\", \""),
+        features_type,
+        supported.join("\", \""),
```

Comment:
we can use [Itertools::join](https://docs.rs/itertools/latest/itertools/fn.join.html):
```suggestion
        supported_features.iter().join("\", \""),
```

And remove the whole computation of `supported` above.

Reply:
Whoa that's slick! It accepts an `Iterator<Item: Display>` instead of requiring strings like normal `join` does!

That means we could also simplify `unsupported` to just:
```rust
    let unsupported = table_features.iter().filter(|f| !supported_features.contains(*f));
```
(and then `join` consume the iterator directly)

Reply:
yea i suppose they would then just appear in the order they are listed in our static Vecs (not sorting that anymore)